### ISSUE TITLE
Time outs at coordinating node for outgoing replica requests

### ISSUE THREAD

**Issue by vigyasharma** [elastic/elasticsearch#29531]
**Describe the feature**: 
Currently replica requests have ```force_execution``` enabled, wherein they override thread pool queue sizes. The request increases queue size and adds itself to it. If the node processing these requests is slow, the queue grows indefinitely. 

Additionally, it seems that request payload is also at the coordinating node (which received the original request from client), until all replica requests have completed. If there is a **single slow shard** in cluster, it keeps accumulating replica requests and coordinating nodes keep waiting on it indefinitely. This eventually increases memory consumption at coordinating node causing it to crash.
 
Example, for a ```_bulk``` request that hits coordinating node, it is divided into sub-requests sent to each primary shard involved. Say primary is on a good node, but replica is on a bad (slow) node; different from coordinating node. After primary passes, coordinating node starts a replica request and waits for slow node to complete. If slow node has disk issues, this can cause an indefinite wait. Occupying memory at coordinating node causing it to crash (along with possibly the slow node that was anyway queueing requests indefinitely)

How does elasticsearch protect itself against such scenarios? Can we time out replica requests at coordinating nodes (and possibly the slow node) to free up memory after some time?
From what I understand, this can lead to inconsistent replicas. The async replica requests are replica's mechanism for syncing with primary (apart from recovering a completely new replica shard). 

1. Can this be handled instead, with replicas pulling a delta of changes from primary asynchronously, without the need for coordinating nodes to keep replica request data.
2. Or can the coordinating node discard its payload once it has an ack that replica request has been added to replica shard queue?
3. Another option is to have a time out on replica requests, and consider shard failure when number of timed out requests breach a threshold.

**Elasticsearch version** <= 5.5

**JVM version** (`java -version`): 1.8

**OS version** (`uname -a` if on a Unix-like system): Linux

**Comment by bleskes** [comment]
Thanks @vidarkongsli it seem you have a good understanding of the system, so I'll just answer your questions at the end. 

> Can this be handled instead, with replicas pulling a delta of changes from primary asynchronously, without the need for coordinating nodes to keep replica request data.

Is it could, but to be effective it has to fundamentally change how Elasticsearch works. It has implications to reads and persistency  - when an indexing operations returns you will have no guarantee it's persistent on all replicas nor reporting of how many replicas it was successful on. I don't see this happening any time soon.

> Or can the coordinating node discard its payload once it has an ack that replica request has been added to replica shard queue?

In order to be resilient to failures of the node holding the primary, it can't. In such a case it needs to be able to resend the request. To do so it needs to keep it until the primary acks it.

> Another option is to have a time out on replica requests, and consider shard failure when number of timed out requests breach a threshold.

This is the valid direction in which we are heading. That said, we currently have no plans on how and when to add some protections. It's tricky - fail too early and you reduced storage/search availability. Fail too late and you slow down indexing.

I'm closing this as won't fix, for now.

**Comment by vigyasharma** [comment]
Thanks @bleskes 
Elaborating more on point 3, (with respect to only writes) what if the coordinating node checks all primary and replica queues before sending a bulk sub-request, and rejects the sub-request itself if these queues are full? Won't have diverging replicas if the sub-request is rejected before primaries are written. 

Regarding read on replicas or primaries, I thought they could just honor client configured timeouts.

Understand the implementation complexities involved; want to understand if I'm conceptually missing something here?

**Comment by bleskes** [comment]
> Elaborating more on point 3, (with respect to only writes) what if the coordinating node checks all primary and replica queues before sending a bulk sub-request, and rejects the sub-request itself if these queues are full? Won't have diverging replicas if the sub-request is rejected before primaries are written.

I think that the gist of what you're saying is that we should somehow gate operations before sending the to the primaries, if we know that one or more of it's replicas is busy. Putting the overhead aside, I see that this will cause part of the bulks to be failed quickly and such that other parts of the bulk can proceed. What worries me in this construct is that it ends up being the same to the client - it now has rejections it needs to deal with instead of just waiting on a response from ES.

> Regarding read on replicas or primaries, I thought they could just honor client configured timeouts.

Read timeouts are different as there is a clear fallback. I suggest you open a new issue to discuss read timeouts. Note though that that comes with it's own bag of complexity.

Can you say more about your setting and the problems you are trying to solve? we can talk more about the general replication setup in ES but concrete examples help tremendously in finding simpler solutions.

**Comment by vigyasharma** [comment]
Hi @bleskes, 

Adding an example of what we observed with some follow up questions. Hope it provides more context for the problem we're trying to solve.

We had a 20 data node. 3 dedicated master node cluster serving production data, and 1 of the 20 data nodes had bad EBS, resulting in very slow writes on that node. This cluster was receiving bulk write traffic comprising mostly of index operations. The slow node was still part of the cluster, all shards on it were active, cluster was green, but bulk thread pool had a lot of queued up requests on that node. We were also using the sonian jetty plugin with synchronous i/o to elasticsearch. It showed us that single bulk write calls were taking up to 30 mins; which kept all jetty threads busy (due to sync i/o in jetty) and starved all other incoming read requests.
**_The issue got resolved as soon as we took that one slow node out of the cluster._**

As a fix going forward, we are moving to async i/o in jetty to begin with. But we want to understand how a single slow node could slow down bulk writes across the cluster, and how we can prevent it.

 
To repro the issue, we simulated on another cluster with 3 nodes, where one node had slow writes to disk. It had one index with replication set to 1 (so 1 primary and 1 replica per shard). We then fired bulk requests on the cluster comprising entirely of index operations.
We noticed that once bulk queue is full, primary shard requests on the slow node get rejected with a 429, but replica shard writes on the node keep waiting. Also it seems that replica requests have unbounded queues on thread pool (force_execution) and indefinite timeout (timeout set to null). To investigate further, we moved shards around so that the slow node had only replica shards now. All requests now seemed to wait on the replica shard before returning and we saw high latency in responses. Bulk queue on slow node also kept increasing in size beyond the configured value for bulk queue.
 _We did not touch the setting for wait_for_active_shards, it was set to default i.e. 1._
 
#### Questions --
 

1. If a write succeeds on primary, and replica shards are still active, will the request wait for all replicas to succeed before returning? Does wait_for_active_shards have a role to play here, or is it only checked before writing to primary?
2. Can we time out this request somehow? Esp. after write to primary has succeeded but it waiting on replica?


We are focusing on primary passing and replica not passing case, because when primary itself is occupied, we've seen elasticsearch reject requests with 429. But if primary succeeds and replica writes are slow, we're concerned that request payload might stay in memory till replica passes. This can accumulate overtime (if replica is really slow and there is high traffic) causing OOM issues. It would be a concern even if after we move to async i/o in jetty or native netty in elasticsearch.

**Comment by bleskes** [comment]
@vigyasharma thanks for explaining. It sounds like everything works as it should given the current trade offs in Elasticsearch. Elasticsearch does synchronous replication and waits on all active replicas of the relevant shard to index a document before acking the write. This is important to allow us to read from any single copy of the data. I'm not saying there aren't any options here but this is how ES currently works.  I'm happy to discuss and explain more but I would ask you to open a thread on our [discuss forums](http://discuss.elastic.co) as we keep github for concrete issues and feature request.
